experimental/grpc_attestation_client/rust/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+                yield message;
+            }
+        };
+
+        let response = client
+            .get_attested_data(Request::new(request_stream))
+            .await
+            .context("Couldn't send request")?;
+        let response_stream = response.into_inner();
+
+        Ok(Self {
+            sender,
+            response_stream,
+        })
+    }
+
+    async fn send(
+        &mut self,
+        request: GetAttestedDataRequest,
+    ) -> anyhow::Result<Option<GetAttestedDataResponse>> {
+        self.sender
+            .send(request)
+            .await
+            .context("Couldn't send request")?;
+        self.response_stream
+            .message()
+            .await
+            .context("Couldn't receive response")
+    }
+}
+
+/// gRPC Attestation Service client implementation.
+pub struct AttestationClient {
+    client: Client,
+    encryptor: AeadEncryptor,
+}
+
+impl AttestationClient {
+    pub async fn create(uri: &str, expected_tee_measurement: &[u8]) -> anyhow::Result<Self> {
+        let mut client = Client::create(uri)
+            .await
+            .context("Couldn't create gRPC client")?;
+        let encryptor = AttestationClient::attest(&mut client, expected_tee_measurement)
+            .await
+            .context("Couldn't attest server")?;
+
+        Ok(Self { client, encryptor })
+    }
+
+    /// Attests server for a single gRPC streaming request.
+    async fn attest(
+        client: &mut Client,
+        expected_tee_measurement: &[u8],
+    ) -> anyhow::Result<AeadEncryptor> {
+        let key_negotiator =
+            DiffieHellmanNegotiator::create().expect("Couldn't create key negotiator");
+        let public_key = key_negotiator
+            .public_key()
+            .expect("Couldn't get public key");
+
+        // Create attestation request with client's public key.
+        let attestation_request = GetAttestedDataRequest {
+            request_type: Some(RequestType::AttestationRequest(AttestationRequest {
+                public_key,
+            })),
+        };
+        let response = client
+            .send(attestation_request)
+            .await
+            .context("Couldn't send attestation request")?
+            .context("Stream stopped preemptively")?;
+
+        // Receive server's attestation containing public key.
+        let response_type = response
+            .response_type
+            .context("Couldn't read response type")?;
+        let attestation_response =
+            if let ResponseType::AttestationResponse(attestation_response) = response_type {
+                Ok(attestation_response)
+            } else {
+                Err(anyhow!("Received incorrect message type"))
+            }?;
+
+        // Verify server's attestation info.
+        AttestationClient::verify_attestation(
+            attestation_response.attestation_info,
+            expected_tee_measurement,
+        )
+        .context("Couldn't verify server attestation")?;
```

Comment:
I think this could be encapsulated in an `AssertionVerifier` or `ClaimVerifier` trait and corresponding struct, of which the TEE measurement one is just a special case.

Reply:
But the method in this struct will still need to have `expected_tee_measurement` as a parameter.
Because making it a `T` parameter may not be enough for functions that require more parameters.

Reply:
Will be refactored as part of the https://github.com/project-oak/oak/issues/2106